kernel/src/snapshot.rs

```diff
@@ -67,6 +76,167 @@ impl Snapshot {
         Self::try_new_from_log_segment(table_root, log_segment, engine)
     }
 
+    /// Create a new [`Snapshot`] instance from an existing [`Snapshot`]. This is useful when you
+    /// already have a [`Snapshot`] lying around and want to do the minimal work to 'update' the
+    /// snapshot to a later version.
+    ///
+    /// We implement a simple heuristic:
+    /// 1. if the new version == existing version, just return the existing snapshot
+    /// 2. if the new version < existing version, error: there is no optimization to do here
+    /// 3. list from (existing checkpoint version + 1) onward (or just existing snapshot version if
+    ///    no checkpoint)
+    /// 4. a. if new checkpoint is found: just create a new snapshot from that checkpoint (and
+    ///    commits after it)
+    ///    b. if no new checkpoint is found: do lightweight P+M replay on the latest commits (after
+    ///    ensuring we only retain commits > any checkpoints)
+    ///
+    /// # Parameters
+    ///
+    /// - `existing_snapshot`: reference to an existing [`Snapshot`]
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `version`: target version of the [`Snapshot`]. None will create a snapshot at the latest
+    ///   version of the table.
+    pub fn try_new_from(
+        existing_snapshot: Arc<Snapshot>,
+        engine: &dyn Engine,
+        version: impl Into<Option<Version>>,
+    ) -> DeltaResult<Arc<Self>> {
+        let old_version = existing_snapshot.version();
+        let new_version = version.into();
+        if let Some(new_version) = new_version {
+            if new_version == old_version {
+                // Re-requesting the same version
+                return Ok(existing_snapshot.clone());
+            }
+            if new_version < old_version {
+                // Hint is too new: error since this is effectively an incorrect optimization
+                return Err(Error::Generic(format!(
+                    "Requested snapshot version {} is older than snapshot hint version {}",
+                    new_version, old_version
+                )));
+            }
+        }
+
+        let log_root = existing_snapshot.log_segment.log_root.clone();
```

Comment:
```suggestion
        let old_log_segment = &existing_snapshot.log_segment;
        let log_root = old_log_segment.log_root.clone();
```
(multiple uses of `existing_snapshot.log_segment`)